rust/otap-dataflow/crates/otap/src/attribute_processor.rs

```diff
@@ -0,0 +1,882 @@
+//!
+//! Unsupported actions are ignored if present in the config:
+//! `insert`, `upsert`, `update` (value update), `hash`, `extract`, `convert`.
+//! We may add support for them later.
+//!
+//! Example configuration (YAML):
+//! You can optionally scope the transformation using `apply_to`. Valid values: signal, resource, scope.
+//! If omitted, defaults to [signal].
+//! ```yaml
+//! actions:
+//!   - action: "rename"
+//!     from: "http.method"
+//!     to: "rpc.method"       # Renames http.method to rpc.method
+//!   - key: "db.statement"
+//!     action: "delete"       # Removes db.statement attribute
+//!   # apply_to: ["signal", "resource"]  # Optional; defaults to ["signal"]
+//! ```
+//!
+//! Implementation uses otel_arrow_rust::otap::transform::transform_attributes for
+//! efficient batch processing of Arrow record batches.
+
+use crate::{OTAP_PROCESSOR_FACTORIES, pdata::OtapPdata};
+use async_trait::async_trait;
+use linkme::distributed_slice;
+use otap_df_config::error::Error as ConfigError;
+use otap_df_config::experimental::SignalType;
+use otap_df_config::node::NodeUserConfig;
+use otap_df_engine::config::ProcessorConfig;
+use otap_df_engine::error::Error as EngineError;
+use otap_df_engine::local::processor as local;
+use otap_df_engine::message::Message;
+use otap_df_engine::processor::ProcessorWrapper;
+use otel_arrow_rust::otap::{
+    OtapArrowRecords,
+    transform::{AttributesTransform, transform_attributes},
+};
+use otel_arrow_rust::proto::opentelemetry::arrow::v1::ArrowPayloadType;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::{BTreeMap, BTreeSet, HashSet};
+use std::sync::Arc;
+
+/// URN for the AttributeProcessor
+pub const ATTRIBUTE_PROCESSOR_URN: &str = "urn:otap:processor:attribute_processor";
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// Actions that can be performed on attributes.
+#[serde(tag = "action", rename_all = "lowercase")]
+pub enum Action {
+    /// Rename an existing attribute key (non-standard; deviates from Go config).
+    Rename {
+        /// The source key to rename from.
+        from: String,
+        /// The destination key to rename to.
+        to: String,
+    },
+
+    /// Delete an attribute by key.
+    Delete {
+        /// The attribute key to delete.
+        key: String,
+    },
+
+    /// Other actions are accepted for forward-compatibility but ignored.
+    /// These variants allow deserialization of Go-style configs without effect.
+    #[serde(other)]
+    Unsupported,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+/// Configuration for the AttributeProcessor.
+///
+/// Accepts configuration in the same format as the OpenTelemetry Collector's attributes processor.
+/// Supported actions: rename (deviation), delete. Others are ignored.
+///
+/// You can control which attribute domains are transformed via `apply_to`.
+/// Valid values: "signal" (default), "resource", "scope".
```

Comment:
I take it signal refers to a signal-specific set of items. This confused me for a second, and I wonder if using the specific signal name would help. Instead of a list like ["signal"] you could use a list like ["logs", "metrics"].